pkg/engine/task/task_pipe.go

```diff
@@ -0,0 +1,407 @@
+	pipePodContainerName = "waiter"
+
+	PipePodAnnotation = "kudo.dev/pipepod"
+)
+
+const (
+	// Max file size of a pipe file.
+	maxPipeFileSize = 1024 * 1024
+)
+
+type PipeFileKind string
+
+const (
+	// PipeFile will be persisted as a Secret
+	PipeFileKindSecret PipeFileKind = "Secret"
+	// PipeFile will be persisted as a ConfigMap
+	PipeFileKindConfigMap PipeFileKind = "ConfigMap"
+)
+
+type PipeTask struct {
+	Name      string
+	Pod       string
+	PipeFiles []PipeFile
+}
+
+type PipeFile struct {
+	File string
+	Kind PipeFileKind
+	Key  string
+}
+
+func (pt PipeTask) Run(ctx Context) (bool, error) {
+	// 1. - Render container template -
+	rendered, err := render([]string{pt.Pod, pt.Pod}, ctx)
+	if err != nil {
+		return false, fatalExecutionError(err, taskRenderingError, ctx.Meta)
+	}
+
+	// 2. - Create core/v1 container object -
+	usrPod, err := unmarshal(rendered[pt.Pod])
+	if err != nil {
+		return false, fatalExecutionError(err, resourceUnmarshalError, ctx.Meta)
+	}
+
+	// 3. - Validate the container object -
+	err = validate(usrPod, pt.PipeFiles)
+	if err != nil {
+		return false, fatalExecutionError(err, resourceValidationError, ctx.Meta)
+	}
+
+	// 4. - Create a pod using the container -
+	podName := PipePodName(ctx.Meta)
+	podYaml, err := pipePod(usrPod, podName)
+	if err != nil {
+		return false, fatalExecutionError(err, pipeTaskError, ctx.Meta)
+	}
+
+	// 5. - Kustomize pod with metadata
+	podObj, err := kustomize(map[string]string{"pipe-pod.yaml": podYaml}, ctx.Meta, ctx.Enhancer)
+	if err != nil {
+		return false, fatalExecutionError(err, taskEnhancementError, ctx.Meta)
+	}
+
+	// 6. - Apply pod using the client -
+	podObj, err = apply(podObj, ctx.Client)
+	if err != nil {
+		return false, err
+	}
+
+	// 7. - Wait for the pod to be ready -
+	err = isHealthy(podObj)
+	// once the pod is Ready, it means that its initContainer finished successfully and we can copy
+	// out the generated files. An error during a health check is not treated as task execution error
+	if err != nil {
+		return false, nil
+	}
+
+	// 8. - Copy out the pipe files -
+	log.Printf("PipeTask: %s/%s copying pipe files", ctx.Meta.InstanceNamespace, ctx.Meta.InstanceName)
+	fs := afero.NewMemMapFs()
+	pipePod := podObj[0].(*corev1.Pod)
+
+	err = copyFiles(fs, pt.PipeFiles, pipePod, ctx)
+	if err != nil {
+		return false, err
+	}
+
+	// 9. - Create k8s artifacts (ConfigMap/Secret) from the pipe files -
+	log.Printf("PipeTask: %s/%s creating pipe artifacts", ctx.Meta.InstanceNamespace, ctx.Meta.InstanceName)
+	artStr, err := storeArtifacts(fs, pt.PipeFiles, ctx.Meta)
+	if err != nil {
+		return false, err
+	}
+
+	// 10. - Kustomize artifacts -
+	artObj, err := kustomize(artStr, ctx.Meta, ctx.Enhancer)
+	if err != nil {
+		return false, fatalExecutionError(err, taskEnhancementError, ctx.Meta)
+	}
+
+	// 11. - Apply artifacts using the client -
+	_, err = apply(artObj, ctx.Client)
+	if err != nil {
+		return false, err
+	}
+
+	// 12. - Delete pipe pod -
```

Comment:
I.e. if any of the steps after applying the pod failed, the pod won't get deleted? Is this expected? Or would deferring the deletion after step 6 make more sense here?

Reply:
No, the task will be re-run as it always happens anyway. All tasks are supposed to idempotent.